feat: Add extended explain info to Comet plan

parthchandra opened the pull request:
## Which issue does this PR close?

Closes #253

## Rationale for this change

Adds addition planning information to a Spark plan which allows us to track the reasons why a Spark plan was not fully converted to Comet.

## What changes are included in this PR?

This PR adds a `CometExplainInfo` structure that is returned by every step of the planning process. Eventually the Structure is attached to the Spark plan being returned by adding it as a Comet specific tag

## How are these changes tested?

Additional unit test and tested with all TPCH and TPCDS queries with both Spark 3.4 and Spark 3.4 with the extended plan support.
Also regenerated the `CometTPC*QueriesList` output for both TPCH and TPCDS run against a 1TB dataset.
Note: Since the extended plan support does not exist before Spark 4, the output of `CometTPC*QueriesList` was updated with a build of Spark 3.4 with the feature backported.

andygrove commented on spark/src/main/scala/org/apache/comet/CometExplainInfo.scala:
```diff
@@ -0,0 +1,81 @@
+ * under the License.
+ */
+
+package org.apache.comet
```
Unrelated to this PR, but we should probably be using `org.apache.datafusion.comet` once the repositories get renamed?

parthchandra replied:
Agree. But this won't be the only class needing that change. So perhaps better to wait until then and do one bulk update?

andygrove replied:
Yes, absolutely. The move will likely happen next week, so let's discuss after that.

andygrove commented on spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:
```diff
@@ -90,6 +98,32 @@ class CometSparkSessionExtensions
+          // unsupported parquet data source V2
+          case scanExec: BatchScanExec if scanExec.scan.isInstanceOf[ParquetScan] =>
+            val requiredSchema = scanExec.scan.asInstanceOf[ParquetScan].readDataSchema
+            val info1 = if (isSchemaSupported(requiredSchema)) {
+              CometExplainInfo(s"Schema $requiredSchema is not supported")
+            } else {
+              CometExplainInfo.none
+            }
```
We could use Scala's `Option` type here rather than defining our own `none` constant. For example:

```scala
            val info1 = if (isSchemaSupported(requiredSchema)) {
              Some(CometExplainInfo(s"Schema $requiredSchema is not supported"))
            } else {
              None
            }
```

We would also need to update the call to pass the list of reasons to `opWithInfo` to use `flatten` so that we only pass the valid reasons.

```scala
opWithInfo(scanExec, CometExplainInfo("SCAN", Seq(info1, info2, info3).flatten))
```

parthchandra replied:
Good idea. I may have a cleaner way of doing this thoguh, which would make this unnecessary. Give me a little time since I need to change a lot of places to try out the idea. If that doesn't work, I'll change this to use Scala `Option`

andygrove commented on spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:
```diff
@@ -400,52 +466,98 @@ class CometSparkSessionExtensions
+              CometTakeOrderedAndProjectExec
+                .isSupported(s)
+                ._1 =>
+          // TODO: support offset for Spark 3.4
```
Do we need to file an issue to track this?

parthchandra replied:
This got copied from some other branch so I've removed it. I'm not too sure what else we need to support for `CometTakeOrderedAndProjectExec` but let me find out and log an issue

andygrove commented on spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:
```diff
@@ -538,12 +659,47 @@ class CometSparkSessionExtensions
           // An operator that is not supported by Comet
-          op
+          op match {
+            case b: CometExec => b
+            case b: CometBroadcastExchangeExec => b
+            case b: CometShuffleExchangeExec => b
```
very minor nit: we could combine these if we want to and avoid the extra variable `b`.

```suggestion
            case _: CometExec | _: CometBroadcastExchangeExec | _: CometShuffleExchangeExec => op
```

dbtsai commented on spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala:
```diff
@@ -0,0 +1,85 @@
+      if (s.nonEmpty) {
+        info = info :+ s
+      }
+    })
```
For longer lambda expression, we can ignore parentheses. Why do we have a `map(t => t)`?
```    
sorted.foreach { p => 
      val s = getActualPlan(p).getTagValue(CometExplainInfo.EXTENSION_INFO).map(t => t).getOrElse("")
      if (s.nonEmpty) {
        info = info :+ s
      }
    }
```

parthchandra replied:
changed

dbtsai commented on spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala:
```diff
@@ -0,0 +1,85 @@
+            case _: SparkPlan => traversed.enqueue(getActualPlan(c.asInstanceOf[SparkPlan]))
+            case _ =>
+          }
+          ()
```
ditto, remove the unneeded parentheses.

parthchandra replied:
chnaged

parthchandra commented:
@andygrove I changed the core of the implementation. Instead of setting information in a CometExplainInfo structure and bubbling it up in the plan, I now set the explain information as a spark tag in the plan or expression.  This makes the changes easier to implement as we add support for more operators and expressions. 
Please take a look.

andygrove commented on spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:
```diff
@@ -91,6 +98,30 @@ class CometSparkSessionExtensions
+          // unsupported parquet data source V2
+          case scanExec: BatchScanExec if scanExec.scan.isInstanceOf[ParquetScan] =>
+            val requiredSchema = scanExec.scan.asInstanceOf[ParquetScan].readDataSchema
+            var info1: Option[String] = None
+            if (isSchemaSupported(requiredSchema)) {
+              info1 = Some(s"Schema $requiredSchema is not supported")
+            }
```
I wonder if we could add a helper method to reduce the boilerplate for each of these sections for creating the `Option[String]`.

We could add a method like this:

```scala
def createMessage(condition: Boolean, message: () => String): Option[String] = {
  if (condition) {
    Some(message())
  } else {
    None
  }
}
```

The call site could then be a one-liner:

```scala
val info1 = createMessage(isSchemaSupported(requiredSchema), s"Schema $requiredSchema is not supported")
```

What do you think?

andygrove replied:
edit: I just updated the suggested code to use the pass-by-name `message: => String` syntax so that the message is evaluated lazily

parthchandra replied:
Done.

parthchandra commented on spark/inspections/CometTPCHQueriesList-results.txt:
```diff
@@ -0,0 +1,121 @@
+Query: q2 TPCH Snappy. Comet Exec: Enabled (CometSort, CometSortMergeJoin, CometProject, CometFilter)
+Query: q2 TPCH Snappy: ExplainInfo:
+ObjectHashAggregate is not supported
+might_contain is not supported
```
Great catch! I was catching this but in FilterExec I was not propagating the correct error. Fixed.

parthchandra commented on spark/inspections/CometTPCHQueriesList-results.txt:
```diff
@@ -0,0 +1,121 @@
+ObjectHashAggregate is not supported
+might_contain is not supported
+BroadcastHashJoin is not supported
+SortMergeJoin is not supported
```
Another good catch! I reviewed everything and found a few more paths where the error was not being caught. Fixed.

parthchandra commented on spark/src/test/scala/org/apache/spark/sql/CometTPCQueryListBase.scala:
```diff
@@ -98,6 +105,11 @@ trait CometTPCQueryListBase
         } else {
           out.println(s"Query: $name$nameSuffix. Comet Exec: Disabled")
         }
+        if (supportsExtendedExplainInfo(df.queryExecution)) {
```
Good point. Done

parthchandra commented:
@advancedxy, @andygrove Addressed your comments. I was missing a few paths where the explanation was not being propagated correctly. I've fixed the ones I could find.
Also, updated the results in CometTPC*QueriesList. I had run this previously on a 1TB dataset and not run these after I rebased on main. I've run it again on a 1GB dataset but had to explicitly set some Bloom filter configuration options to simulate the conditions of the larger data set. Otherwise the settings are the same as the defaults we expect most users to have.

andygrove reviewed (approved):
LGTM. Thanks @parthchandra

parthchandra commented:
Looking into the ci failures ...

parthchandra commented:
@andygrove could you please trigger a re-run of the failed ci workflow? (Only committers can do that). I cannot reproduce the failure locally and am pretty sure it has nothing to do with this PR.

viirya commented:
Re-triggered.

parthchandra commented:
Could we merge this please? @viirya @andygrove

viirya commented on spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:
```diff
@@ -128,6 +158,26 @@ class CometSparkSessionExtensions
+              s"Schema $requiredSchema is not supported")
+            val info2 = createMessage(
+              !isSchemaSupported(partitionSchema),
+              s"Schema $partitionSchema is not supported")
```
```suggestion
              s"Partition schema $partitionSchema is not supported")
```

viirya commented on spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala:
```diff
@@ -0,0 +1,87 @@
+    info.distinct.mkString("\n").trim
+  }
+
+  private def getActualPlan(node: TreeNode[_]): TreeNode[_] = {
```
Actually this only works on SparkPlan, I'm wondering why we use `TreeNode[_]` in many places?

viirya commented on spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala:
```diff
@@ -0,0 +1,87 @@
+    info.filter(!_.contentEquals("\n"))
+  }
+
+  // get all plan nodes, breadth first, leaf nodes first
```
breadth first and leaf nodes first seem conflicting?

viirya commented on spark/src/main/scala/org/apache/comet/shims/ShimCometSparkSessionExtensions.scala:
```diff
@@ -49,4 +49,17 @@ object ShimCometSparkSessionExtensions {
+
+  def supportsExtendedExplainInfo(qe: QueryExecution): Boolean = {
+    try {
+      // Look for QueryExecution.extendedExplainInfo(scala.Function1[String, Unit], SparkPlan)
```
Can we add clear reference here that when it is added to Spark, e.g., Spark xx+?

viirya commented on spark/src/main/scala/org/apache/spark/sql/ExtendedExplainGenerator.scala:
```diff
@@ -0,0 +1,32 @@
+
+import org.apache.spark.sql.execution.SparkPlan
+
+/**
+ * A trait for a session extension to implement that provides addition explain plan information.
+ */
+
```
```suggestion
/**
 * A trait for a session extension to implement that provides addition explain plan information.
 * We copy this from Spark 4.0 since this trait is not available in Spark 3.x. We can remove this
 * after dropping Spark 3.x support.
 */
```

viirya commented on spark/src/test/scala/org/apache/spark/sql/CometTPCQueryListBase.scala:
```diff
@@ -84,11 +86,15 @@ trait CometTPCQueryListBase
-        CometConf.COMET_EXEC_ALL_OPERATOR_ENABLED.key -> "true") {
+        CometConf.COMET_EXEC_SHUFFLE_ENABLED.key -> "true",
+        CometConf.COMET_EXEC_ALL_OPERATOR_ENABLED.key -> "true",
+        "spark.sql.optimizer.runtime.bloomFilter.creationSideThreshold" -> "1MB",
+        "spark.sql.optimizer.runtime.bloomFilter.applicationSideScanSizeThreshold" -> "1MB") {
```
Why we need these `bloomFilter` configs?

viirya commented on spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:
```diff
@@ -917,6 +1026,7 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde {
                 .setStringSpace(builder)
                 .build())
           } else {
+            withInfo(expr, null, child)
```
Can we override this method for this kind of usage so we can do `withInfo(expr, child)`?

Can be in a follow up PR.

viirya commented:
There is conflict needed to be resolved. And I have a few minor comments like code comment changes.

viirya commented:
Thank you @parthchandra. I will merge this once CI passes.

parthchandra commented:
Thanks for the review @viirya @andygrove @advancedxy

viirya commented:
Merged. Thanks all.
